### PR TITLE
crystal: try to rebuild for Linuxbrew

### PR DESCRIPTION
**Pull request by iMichka**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

### CODE REVIEW

**Comment by iMichka** [comment]
```
==> make deps
Using /home/linuxbrew/.linuxbrew/opt/llvm/bin/llvm-config [version=10.0.0]
g++-5 -c -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/home/linuxbrew/.linuxbrew/Cellar/llvm/10.0.0_3/include -std=c++14 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
gcc-5 -fPIC -c -o src/ext/sigfault.o src/ext/sigfault.c
ar -rcs src/ext/libcrystal.a src/ext/sigfault.o
==> bin/crystal build -D without_openssl -D without_zlib -D preview_overflow -o .build/crystal src/compiler/crystal.cr --release --no-debug --threads=1
==> Downloading https://github.com/crystal-lang/shards/archive/v0.10.0.tar.gz
Already downloaded: /github/home/.cache/Homebrew/downloads/d2c8e1751029c4f8142629a2f6d6eee801a13cc6bf21bec390495457d2388ae2--shards-0.10.0.tar.gz
==> Verifying d2c8e1751029c4f8142629a2f6d6eee801a13cc6bf21bec390495457d2388ae2--shards-0.10.0.tar.gz checksum
tar xof /github/home/.cache/Homebrew/downloads/d2c8e1751029c4f8142629a2f6d6eee801a13cc6bf21bec390495457d2388ae2--shards-0.10.0.tar.gz -C /tmp/d20200407-19425-1f53efq
cp -pR /tmp/d20200407-19425-1f53efq/shards-0.10.0/. /tmp/crystal--shards-20200407-19425-1w8yhlf/shards-0.10.0
chmod -Rf +w /tmp/d20200407-19425-1f53efq
==> /tmp/crystal-20200407-19425-9fcnzx/crystal-0.34.0/boot/embedded/bin/shards install --production
Failed to execute: /tmp/crystal-20200407-19425-9fcnzx/crystal-0.34.0/boot/embedded/bin/shards

==> Formula
Path: /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/crystal.rb
```

**Comment by bcardiff** [comment]
In the linux packages used for boot, the shards binary is in another location.

The following should do it:

```patch
--- system buildpath/"boot/embedded/bin/shards", "install",
+++ system buildpath/"boot/bin/shards", "install",
```

**Comment by iMichka** [comment]
Thanks. CI failed (it's not the first time with crytsal) with:
```
==> bin/crystal build -D without_openssl -D without_zlib -D preview_overflow -o .build/crystal src/compiler/crystal.cr --release --no-debug --threads=1
Invalid memory access (signal 11) at address 0x7fe2833fed18
[0x7fe2d40d7ee6] ???
[0x7fe2d402ea6b] ???
[0x7fe2d4ea5491] ???
```

This is unrelated to the shards change I just made. I have already seen this error, and I think it happens randomly during the build. I'll re-trigger a new build.

**Comment by bcardiff** [comment]
Noted. I’ve seen that too in our CI very very sporadically.

**Comment by dawidd6** [comment]
Test fails with some weird errors:

```
[0x7f0e987029c9] _ZN4llvm16DwarfCompileUnit37applySubprogramAttributesToDefinitionEPKNS_12DISubprogramERNS_3DIEE +121
[0x7f0e98718a12] _ZN4llvm10DwarfDebug27finishSubprogramDefinitionsEv +82
[0x7f0e98718a7a] _ZN4llvm10DwarfDebug18finalizeModuleInfoEv +42
[0x7f0e98721bd0] _ZN4llvm10DwarfDebug9endModuleEv +80
[0x7f0e986d9fef] _ZN4llvm10AsmPrinter14doFinalizationERNS_6ModuleE +575
[0x7f0e98067a81] _ZN4llvm13FPPassManager14doFinalizationERNS_6ModuleE +97
[0x7f0e980731ec] _ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE +1212
[0x7f0e996e3f8f] ???
[0x7f0e996e41e4] LLVMTargetMachineEmitToFile +212
[0xa0f1bf] ???
[0xa0f9cd] ???
[0x437160] ???
[0x0] ???
gcc: error: _main.o: No such file or directory
gcc: error: S-lice40U-I-nt841.o: No such file or directory
gcc: error: P-ointer40U-I-nt841.o: No such file or directory
gcc: error: A-rgumentE-rror.o: No such file or directory
gcc: error: C-allS-tack.o: No such file or directory
gcc: error: A-rray40P-ointer40V-oid4141.o: No such file or directory
gcc: error: P-ointer40P-ointer40V-oid4141.o: No such file or directory
gcc: error: E-xception.o: No such file or directory
```

**Comment by NIFR91** [comment]
When upgrading in a linux cluster (no root access)
 
```
Linux el-insurgente 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

It always gives me the  mentioned error 

```
==> bin/crystal build -D without_openssl -D without_zlib -D preview_overflow -o .build/crystal src/compiler/crystal.cr --release --no-debug --threads=1
Invalid memory access (signal 11) at address 0x7fe2833fed18
[0x7fe2d40d7ee6] ???
[0x7fe2d402ea6b] ???
[0x7fe2d4ea5491] ???
```

**Comment by ipolyzos** [comment]
I tried  to rebuild for Linuxbrew on Fedora 31  i.e

```
$ uname -a
Linux ostation 5.5.15-200.fc31.x86_64 #1 SMP Thu Apr 2 19:16:17 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
And I have the same issue with @iMichka  i.e

```
Invalid memory access (signal 11) at address 0x7f05dcc21098
[0x7f062dbefee6] ???
[0x7f062db46a6b] ???
[0x7f062e9bd491] ???
```

_P.S you can find the full log at https://pastebin.com/qHABp35X_

**Comment by bcardiff** [comment]
How much ram those build environment have? It usually requires 4-6gb to build the compiler in release mode IIRC.

Another barrier we reached at some point in some environmets is to use lld instead of ld. But this was experienced in CI environment where the bulid generates many more .o files due to building all std-lib and its specs; and not in the build-only-the-compiler scenarios.

**Comment by ipolyzos** [comment]
> How much ram those build environment have? It usually requires 4-6gb to build the compiler in release mode IIRC.
> 
> Another barrier we reached at some point in some environmets is to use lld instead of ld. But this was experienced in CI environment where the bulid generates many more .o files due to building all std-lib and its specs; and not in the build-only-the-compiler scenarios.

@bcardiff my machine has 64GB

**Comment by iMichka** [comment]
It built reliably before. One of the solutions would be to run a `git bisect` between crystal 0.33 and 0.34 and try to find out when the issue was introduced.

Could somebody also open a issue upstream?

**Comment by dawidd6** [comment]
@bcardiff Any

File: Formula/crystal.rb [hunk 1]
```diff
@@ -96,8 +96,8 @@ def install
 
     # Build shards
     resource("shards").stage do
-      system buildpath/"boot/embedded/bin/shards", "install",
-                                                   "--production"
+      shard = OS.mac? ? buildpath/"boot/embedded/bin/shards" : buildpath/"boot/bin/shards"
+      system shard, "install", "--production"
 
       system buildpath/"bin/crystal", "build",
                                       "-o", buildpath/".build/shards",
```
Comment (bcardiff) [inline Formula/crystal.rb:103]: @iMichka can you check if using `--threads=1` helps the build?

```suggestion
                                      "-o", buildpath/".build/shards",
                                      "--threads=1"
```

Comment (bcardiff) [reply Formula/crystal.rb:103]: hmmm, the paste bien shows `--threads=1` is been used, yet it's not stated here 🤔

Comment (iMichka) [reply Formula/crystal.rb]: I started a new build with this flag.

**Comment by iMichka** [comment]
Looks like nobody is working on this. An issue has been opened upstream. Let's close this, and wait for feedback.

**Pull request closed without merging**